numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -795,8 +811,33 @@ PyArray_CanCastTypeTo(PyArray_Descr *from, PyArray_Descr *to,
     }
     /* If safe or same-kind casts are allowed */
     else if (casting == NPY_SAFE_CASTING || casting == NPY_SAME_KIND_CASTING) {
-        if (PyArray_CanCastTo(from, to)) {
-            return 1;
+        /*
+         * Casting from simple to/from types with fields is only OK
+         * if there is only a single field and casting to/from that
+         * field works.
+         */
+        if (PyDataType_HASFIELDS(from)) {
+            PyArray_Descr *only_field = get_only_field(from->fields);
+            /* to cannot have fields, as that would mean EquivTypenum */
+            assert(!PyDataType_HASFIELDS(to));
+            if (only_field != NULL) {
+                return PyArray_CanCastTypeTo(only_field, to, casting);
+            }
+            else {
+                return 0;
+            }
+        }
+        else if (PyDataType_HASFIELDS(to)) {
+            PyArray_Descr *only_field = get_only_field(to->fields);
+            if (only_field != NULL) {
+                return PyArray_CanCastTypeTo(from, only_field, casting);
+            }
+            else {
+                return 0;
+            }
```

Comment:
I could perhaps see an argument that this should be `any(PyArray_CanCastTypeTo(from, field, casting) for field in fields)`, allowing `2` to cast into `(2, 2, 2,)`, which I think we have precedent for

Reply:
I was perhaps more driven by symmetry here than anything else. But I think you are right: casting to multiple fields is in principle not ambiguous. But it should be `all(cancast(...))`, then, not `any(...)`, correct?

Reply:
To take a step back here, do we want safe casting? It feels right, but it also feels right that safe casting in most cases aligns with "`common_type`" and I am not sure I feel that I feel things like `np.concatenate` should feel they can just "broadcast" to all fields without some manual casting. (I did not make up my mind yet, probably its OK).

Reply:
To me it's a question of if we already set a precedent for this weird behavior. If we didn't, then the original more conservative patch was probably the right one.

Reply:
I'm moving this out of the in-line comments, so it doesn't get lost when I change things.